README.md
# LudumDare_52

Source code of a Pico-8 game made in 48 hours for the Ludum Dare 52

The theme of the game jam was "Harvest"

links to the uploaded game : https://mastoast.itch.io/youre-under-harvest

CHANGELOG.md
# changelog
All notable changes to this project will be documented in this file.

## 1.0.0 - Jam version - 08/01/2023
 - Initial release

## 1.1.0 - Post jam - update from feedback - 11/01/2023
 - Added "level restart" and "back to menu" options from the pause menu during a level
 - Added jump buffer to help jumping multiple times in a short period of time
 - Added persistent data storage for level scorings
 - Added check of old scores to prevent overwriting an existing score with a lesser score
 - Reworked plants sprites to help pressing corresponding keys
 - Reworked music of level 1
 - Reworked music of level 2
 - Reworked menu navigation bindings
 - Reworked plants particles
 - Reworked collection validation ui elements
 - Added Changelog (hi there)

## 1.2.0 - 13/01/2023
 - Added an extra level with a 4th direction to press
 - Added visual rewards on perfect score

levels.lua
level_0 = {}
level_0.name = "Tutorial"
level_0.start_music = 0
level_0.ch_mask = 7
level_0.speed = 16
level_0.events = {{x = 0, y = 18}, {x = 0, y = 19}}
level_0.event_offset = 64
-- player best
level_0.best_score = -1
-- maximum possible score on level
level_0.max_score = -1
level_0.saved_score = -1

level_1 = {}
level_1.name = "Level 1"
level_1.start_music = 10
level_1.ch_mask = 7
level_1.speed = 16
level_1.events = {{x = 0, y = 22}, {x = 0, y = 23}}
level_1.event_offset = 64
level_1.best_score = -1
level_1.max_score = -1
level_1.saved_score = -1

level_2 = {}
level_2.name = "Level 2"
level_2.start_music = 18
level_2.ch_mask = 7
level_2.speed = 16
level_2.events = {{x = 0, y = 27}, {x = 0, y = 28}, {x = 0, y = 29}}
level_2.event_offset = 64
level_2.best_score = -1
level_2.max_score = -1
level_2.saved_score = -1

level_3 = {}
level_3.name = "Level 3"
level_3.start_music = 28
level_3.ch_mask = 7
level_3.speed = 16
level_3.events = {{x = 0, y = 32}, {x = 0, y = 33}, {x = 0, y = 34}, {x = 0, y = 35}}
level_3.event_offset = 64
level_3.best_score = -1
level_3.max_score = -1
level_3.saved_score = -1

function init_level(level)
    current_level = level
    debug_pattern_offset = debug_pattern_offset or 0
    gtime = 0
    gstate = 1
    combo_count = 0
    best_combo = 0
    score_count = 0
    final_score = -1
    objects = {}
    particles = {}
    ground_limit = 96
    create(moon, 95, 25)
    --
    menuitem(1, "Restart level", function() init_level(current_level) end)
    menuitem(2, "Back to menu", function() init_menu() end)
    new_highscore = false
    --
    current_runner = create(runner, 24 + debug_pattern_offset*8*32, 88)
    local title = create(textc, 64, 56)
    title.text = level.name
    title.life = 320

    -- tutorial
    tuto1 = create(text, 20, 20)
    tuto1.text = "⬆️ => "
    tuto1.spr = apple.spr
    tuto1.life = 510
    tuto2 = create(text, 20, 30)
    tuto2.text = "⬅️ => "
    tuto2.spr = melon.spr
    tuto2.life = 520
    tuto3 = create(text, 20, 40)
    tuto3.text = "➡️ => "
    tuto3.spr = leak.spr
    tuto3.life = 530
    if level == level_3 then
        tuto4 = create(text, 20, 50)
        tuto4.text = "⬇️ => "
        tuto4.spr = carrot.spr
        tuto4.life = 530
    end

    -- gen objects from map
    local event_list_count = 0
    local pixels_per_row = 8
    for event_list in all(level.events) do
        local y = event_list.y
        local x_offset = level.event_offset*8 + event_list_count*8*128
        for i=event_list.x, 127 do
            if mget(i, y) == leak.spr then
                create(leak, i*pixels_per_row + x_offset, ground_limit - leak.hit_h)
            elseif mget(i, y) == melon.spr then
                create(melon, i*pixels_per_row + x_offset, ground_limit - melon.hit_h)
            elseif mget(i, y) == apple.spr then
                create(apple, i*pixels_per_row + x_offset, ground_limit - 2*8 - apple.hit_h)
            elseif mget(i, y) == carrot.spr then
                create(carrot, i*pixels_per_row + x_offset, ground_limit - carrot.hit_h)
            end
        end
        event_list_count += 1
    end
    sfx(1, 3, 0, 8)
    music(level.start_music + debug_pattern_offset, 0, level.ch_mask)
end

object.lua
objects = {}
types = {}
lookup = {}
function lookup.__index(self, i) return self.base[i] end

object = {}
object.speed_x = 0;
object.speed_y = 0;
object.remainder_x = 0;
object.remainder_y = 0;
object.hit_x = 0
object.hit_y = 0
object.hit_w = 8
object.hit_h = 8
object.facing = 1
object.solid = true
object.freeze = 0

function object.init(self) end
function object.update(self) end
function object.draw(self)
    spr(self.spr, self.x, self.y, 1, 1, self.flip_x, self.flip_y)
end

function object.move_x(self, x, on_collide)
    self.remainder_x += x
    local mx = flr(self.remainder_x + 0.5)
    self.remainder_x -= mx

    local total = mx
    local mxs = sgn(mx)
    while mx != 0
    do
        if self:check_solid(mxs, 0) then
            if on_collide then
                on_collide(self, total - mx, total)
            end
            return true
        else
            self.x += mxs
            mx -= mxs
        end
    end

    return false
end

function object.move_y(self, y, on_collide)
    self.remainder_y += y
    local my = flr(self.remainder_y + 0.5)
    self.remainder_y -= my

    local total = my
    local mys = sgn(my)
    while my != 0
    do
        if self:check_solid(0, mys) then
            if on_collide then
                on_collide(self, total - my, total)
            end
            return true
        else
            self.y += mys
            my -= mys
        end
    end
    
    return false
end

function object.overlaps(self, b, ox, oy)
    if self == b then return false end
    ox = ox or 0
    oy = oy or 0
    return
        ox + self.x + self.hit_x + self.hit_w > b.x + b.hit_x and
        oy + self.y + self.hit_y + self.hit_h > b.y + b.hit_y and
        ox + self.x + self.hit_x < b.x + b.hit_x + b.hit_w and
        oy + self.y + self.hit_y < b.y + b.hit_y + b.hit_h
end

function object.contains(self, px, py)
    return
        px >= self.x + self.hit_x and
        px < self.x + self.hit_x + self.hit_w and
        py >= self.y + self.hit_y and
        py < self.y + self.hit_y + self.hit_h
end

function object.check_solid(self, ox, oy)
    ox = ox or 0
    oy = oy or 0

    if oy + self.y + self.hit_h > ground_limit then return true end

    -- object collisions
    if not self.ghost then
        for o in all(objects) do
            if o != self and not o.destroyed and o.solid and self:overlaps(o, ox, oy) then
                return true
            end
        end
    end

end

function create(type, x, y, hit_w, hit_h)
    local obj = {}
    obj.base = type
    obj.x = x
    obj.y = y
    obj.hit_w = hit_w or type.hit_w or 8
    obj.hit_h = hit_h or type.hit_h or 8
    setmetatable(obj, lookup)
    add(objects, obj)
    obj:init()
    return obj
end

function new_type(spr)
    local obj = {}
    obj.spr = spr
    obj.base = object
    setmetatable(obj, lookup)
    types[spr] = obj
    return obj
end


runner.lua
runner = new_type(4)
runner.hit_w = 8
runner.hit_h = 8
runner.sprs = {4, 5}
runner.anim_time = 7
runner.flr = true
runner.jump_force = -2.9
runner.gravity = 0.20
runner.action_spr = 6
runner.grid_count = 0
-- validation ui elements distance to the player
runner.marks_y_offset = -8

function runner.init(self)
    self.speed_x = 1
    self.state = 1
    self.end_anim_time = 0
    self.grid_count = 0
    self.buffured_jump = false
end

function runner.update(self)
    local input_y = 0

    -- fix synch every tick
    local delay = 24 + (stat(56)/current_level.speed) * 8 + (debug_pattern_offset + stat(55)) * 256 - self.x
    if delay >= 2 then
        self.x += 2
    elseif delay > 0 then
        self.x += 1
    end
    
    -- fix synch every pattern
    -- if stat(55) != self.grid_count then
    --     self.grid_count += 1
    --     -- 16 ticks for one row
    --     -- 8 pixels per row
    --     -- 256 pixels per pattern
    --     shake = 0
    --     --self.x = 24 + (stat(56)/current_level.speed) * 8 + stat(55) * 256
    -- end
    
    -- anim
    if gtime <= self.end_anim_time then self.spr = self.action_spr
    elseif gtime%self.anim_time == 0 then
        self.spr = self.sprs[(self.spr + 1)%#self.sprs + 1]
    end

    -- grounded
    if self.y < ground_limit - self.hit_h then
        self.flr = false
    else
        self.flr = true
    end
    -- jump
    if self.flr and (btnp(⬆️) or buffured_jump) then
        self.speed_y = self.jump_force
        buffured_jump = false
    end
    -- gravity
    if self.speed_y > 0 then
        self.speed_y += self.gravity * 4
        if btnp(⬆️) then buffured_jump = true end
    else
        self.speed_y += self.gravity
    end

    self:move_y(self.speed_y)

    if gstate == 2 then return end

    -- harvest :)
    -- can optimize

    -- melons
    if btnp(⬅️) then
        local success = false
        self.end_anim_time = gtime + self.anim_time
        for o in all(objects) do
            if (o.base == melon) and not o.collected and self:overlaps(o) then
                o.collected = true
                spawn_particles(3 + rnd(4), 3, o.x, o.y, 3)
                success = true
                break
            end
        end
        if success then
            create(valid_ui, self.x, self.y + self.marks_y_offset)
            score_count += collected_plant_score_count
            incr_combo()
        else
            create(invalid_ui, self.x, self.y + self.marks_y_offset)
            score_count = max(score_count + wrong_input_score_count, 0)
            combo_count = 0
        end
    end

    -- leaks
    if btnp(➡️) then
        local success = false
        self.end_anim_time = gtime + self.anim_time
        for o in all(objects) do
            if (o.base == leak) and not o.collected and self:overlaps(o) then
                o.collected = true
                spawn_particles(3 + rnd(4), 4, o.x, o.y, 4)
                success = true
                break
            end
        end
        if success then
            create(valid_ui, self.x, self.y + self.marks_y_offset)
            score_count += collected_plant_score_count
            incr_combo()
        else
            create(invalid_ui, self.x, self.y + self.marks_y_offset)
            score_count = max(score_count + wrong_input_score_count, 0)
            combo_count = 0
        end
    end

    -- apples
    if not self.flr then
        for o in all(objects) do
            if (o.base == apple) and not o.collected and self:overlaps(o) then
                o.collected = true
                spawn_particles(3 + rnd(3), 4, o.x, o.y, 8)
                create(valid_ui, self.x, self.y + self.marks_y_offset)
                score_count += collected_plant_score_count
                incr_combo()
                break
            end
        end
    end

    -- carrots
    if btnp(⬇️) then
        local success = false
        self.end_anim_time = gtime + self.anim_time
        for o in all(objects) do
            if (o.base == carrot) and not o.collected and self:overlaps(o) then
                o.collected = true
                spawn_particles(3 + rnd(3), 4, o.x, o.y, 9)
                success = true
                printable = gtime%2
                break
            end
        end
        if success then
            create(valid_ui, self.x, self.y + self.marks_y_offset)
            score_count += collected_plant_score_count
            incr_combo()
        else
            create(invalid_ui, self.x, self.y + self.marks_y_offset)
            score_count = max(score_count + wrong_input_score_count, 0)
            combo_count = 0
        end
    end

end

function runner.draw(self)
    spr(self.spr, self.x, self.y)
end

function incr_combo()
    combo_count += 1
    if combo_count >= best_combo then best_combo = combo_count end
end

objects.lua
rectangle = new_type(0)
rectangle.color = 7

function rectangle.draw(self)
    rectfill(self.x, self.y, self.x + self.hit_w - 1, self.y + self.hit_h - 1, self.color)
end

-- FRUITS

function fruit_draw(self)
	if not self.collected then
		spr(self.spr, self.x, self.y, 1, 1)
	else
		spr(self.collected_sprite, self.x, self.y, 1, 1)
	end
end

function fruit_update(self)
	if not self.collected and not self.missed and current_runner.x > self.x + self.hit_w then
		self.missed = true
		create(invalid_ui, self.x, self.y)
		score_count = max(score_count + missed_plant_score_count, 0)
		combo_count = 0
	end
end

-- melon
melon = new_type(53)
melon.solid = false
melon.collected_sprite = 54
melon.draw = fruit_draw
melon.update = fruit_update

-- apple
apple = new_type(21)
apple.solid = false
apple.collected_sprite = 22
apple.update = fruit_update

function apple.draw(self)
	spr(9, self.x - 12, self.y - 8, 3, 4)
	-- local stretch_offset = 0
	-- if (stat(56)/current_level.speed) % 8 > 4 then stretch_offset = 1 end
	-- sspr(72, 0, 3*8, 4*8, self.x - 12, self.y - 8, 3*8, 4*8)
	fruit_draw(self)
end

-- leak
leak = new_type(37)
leak.collected_sprite = 38
leak.solid = false
leak.draw = fruit_draw
leak.update = fruit_update

-- carrot
carrot = new_type(23)
carrot.collected_sprite = 24
carrot.solid = false
carrot.draw = fruit_draw
carrot.update = fruit_update

-- BACKGROUND
moon = new_type(0)
moon.solid = false
moon.hit_h = 12
moon.filler = ♪
moon.rsize_min = 5
moon.rsize_max = 8
moon.rsize = 5

function moon.update(self)
	self.rsize = self.rsize_max - (stat(56)/current_level.speed) % 4
end

function moon.draw(self)
	circfill(cam.x + self.x, cam.y + self.y, self.hit_h, 7)
	fillp(self.filler)
	circ(cam.x + self.x, cam.y + self.y, self.hit_h + self.rsize, 7)
	circ(cam.x + self.x, cam.y + self.y, self.hit_h + self.rsize*2, 7)
	fillp()
end

-- UI
valid_ui = new_type(44)
valid_ui.solid = false
valid_ui.life = 30

function valid_ui.update(self)
	self.y -= 1 * gtime%2
	self.life -= 1
	if self.life == 0 then self.destroyed = true end
end

invalid_ui = new_type(45)
invalid_ui.solid = false
invalid_ui.life = 30
invalid_ui.update = valid_ui.update

-- text with lifetime
text = new_type(0)
text.solid = false
text.life = 500
text.text = ""
text.spr = nil

function text.update(self)
	self.life -= 1
	if self.life <= 0 then self.y -= 1 end
	if self.y < -1 then self.destroyed = true end
end

function text.draw(self)
	print(self.text, cam.x + self.x + 1, cam.y + self.y + 1, 0)
	print(self.text, cam.x + self.x, cam.y + self.y, 7)
	if self.spr then
		spr(self.spr, cam.x + self.x + 2 + #self.text*4, cam.y + self.y - 2)
	end
end

-- texting printing level name at start of the level
textc = new_type(0)
textc.solid = false
textc.text = ""
textc.offset_x = 90

function textc.update(self)
	if self.offset_x < -90 then self.destroyed = true end
	if self.offset_x > 6 or self.offset_x < -6 then self.offset_x -= 2
	elseif self.offset_x > -6 then self.offset_x -= 0.05 end
end

function textc.draw(self)
	print_centered(self.text, self.offset_x, cam.y + self.y + 1, 9)
	print_centered(self.text, -self.offset_x, cam.y + self.y, 7)
end

-- PARTICLES

particles = {}

-- number of particles
-- size
-- x / y
-- color
function spawn_particles(nb,size,x,y,color)
	for i=1,flr(nb) do
        add(particles, make_particle(size,x,y,color))
	end
end

function make_particle(size,x,y,color)
	local p={
		size=size or 1,
		color=color or 7,
		x=x,y=y,k=k,
		t=0, t_max=16+flr(rnd(4)),
		dx=rnd(2)-1,dy=-rnd(3),
		ddy=0.05,
		update=update_particle,
		draw=draw_particle
	}
	return p
end

function draw_particle(a)
	circfill(a.x,a.y,a.size,a.color)
end

function update_particle(a)
	if a.size >= 1 and a.t % 4 == 0 then a.size -=1 end
	if a.t % 2 == 0 then
		a.dy += a.ddy
		a.x += a.dx
		a.y += a.dy
	end
	a.t += 1
	if (a.t == a.t_max) del(particles, a)
end

main.lua
function _init()
    gtime = 0
    gstate = 0 -- 0-menu | 1-level | 2-level ending
    ndeath = 0
    freeze_time = 0
    shake = 0
    cam = {x = 0, y = 0}
    printable = 0
    level_list = {level_0, level_1, level_2, level_3}
    -- if level score > 32 768 it goes negative !!
    collected_plant_score_count = 125
    missed_plant_score_count = -40
    wrong_input_score_count = -20
    --
    selected_level = 0
    new_highscore = false
    init_menu()
    -- DEBUG
    debug_pattern_offset = 0
    --init_level(level_3)

    -- load previous high scores
    cartdata("mastoast_underharvest_v1")
    for index = 1, #level_list do
        level_list[index].saved_score = (dget(index) == 0 and -1) or dget(index) 
    end
end

function init_menu()
    gtime = 0
    gstate = 0
    tutorial_shown = false
    cam.x = 0
    cam.y = 0
    objects = {}
    particles = {}
    -- stop every current sounds
    sfx(-1)
    music(-1)
    -- remove level menu items
    menuitem(1)
    menuitem(2)
end

function _update60()
    -- timers
    gtime += 1

    if gstate == 0 then update_menu()
    else update_level() end
end

function _draw()
    if gstate == 0 then draw_menu()
    else draw_level() end

    --print(printable, cam.x + 80, cam.y + 120, 0)
end

-- MENU
function update_menu()
    if not tutorial_shown then
        if btnp(⬇️) then
            sfx(2, 0, 8, 4)
            if (selected_level+1 == #level_list) then
                tutorial_shown = true
                selected_level = 0
            else
                selected_level = (selected_level + 1)%#level_list
            end
        end
        if btnp(⬆️) then
            sfx(2, 0, 8, 4)
            selected_level = (selected_level - 1)%#level_list
        end
        if btnp(❎) then
            init_level(level_list[selected_level+1])
        end
    else
        if btnp(❎) then
            sfx(2, 0, 8, 4)
            tutorial_shown = false
        end
    end
end

function draw_menu()
    cls(1)
    camera(cam.x, cam.y)
    
    -- background
    line(cam.x - 40, 128, cam.x + 4, 35, 5)
    line(cam.x + 4, 35, cam.x + 68, 128, 5)
    line(cam.x + 30, 74, cam.x + 55, 55, 5)
    line(cam.x + 55, 55, cam.x + 88, 135, 5)
    line(cam.x + 13, 128, cam.x + 150, 65, 5)
    line(cam.x + 150, 65, cam.x + 210, 128, 5)

    map(0, 47, 0, 0, 16, 16, 0)
    spr(apple.spr, 78, 83)
    circfill(cam.x + 95, cam.y + 25, 12, 7)

    if tutorial_shown then
        print_centered("Use the correct button", 1, 4, 0)
        print_centered("Use the correct button", 0, 3, 3)
        print_centered("in rythm to harvest", 1, 12, 0)
        print_centered("in rythm to harvest", 0, 11, 3)
        print_centered("the corresponding item", 1, 20, 0)
        print_centered("the corresponding item", 0, 19, 3)

        print("⬆️ => ", 11, 41, 0)
        print("⬆️ => ", 10, 40, (btn(⬆️) and 8) or 7)
        spr(apple.spr, 35, 39)
        print("⬅️ => ", 11, 51, 0)
        print("⬅️ => ", 10, 50, (btn(⬅️) and 8) or 7)
        spr(melon.spr, 35, 49)
        print("➡️ => ", 11, 61, 0)
        print("➡️ => ", 10, 60, (btn(➡️) and 8) or 7)
        spr(leak.spr, 35, 59)
        print("⬇️ => ", 11, 71, 0)
        print("⬇️ => ", 10, 70, (btn(⬇️) and 8) or 7)
        spr(carrot.spr, 35, 69)

        print("<= TRY IT !", 60, 51, 0)
        print("<= TRY IT !", 60, 50, 8)

        print_centered("❎ back to menu", 1, 116, 0)
        print_centered("❎ back to menu", 0, 115, 7)
    else
        print_centered("You're under harvest", 1, 11, 0)
        print_centered("You're under harvest", 1, 10, 0)
        print_centered("You're under harvest", 0, 10, 3)

        -- level selection
        local incr = 0
        for lvl in all(level_list) do
            local lvl_label = (level_list[selected_level+1] == lvl and "♪ "..lvl.name.." ♪")
            or lvl.name
            print_centered(lvl_label, 1, 41 + 8*incr, 0)
            print_centered(lvl_label, 0, 40 + 8*incr, 7)
            if lvl.saved_score > 0 then
                print(lvl.saved_score.."%", 101, 41 + 8*incr, 8)
                print(lvl.saved_score.."%", 100, 40 + 8*incr, 7)
            end
            incr += 1
        end

        print_centered("⬇️ how to play ⬇️", 1, 76, 0)
        print_centered("⬇️ how to play ⬇️", 0, 75, 7)

        if gtime %128 > 32 then 
            print_centered(" press ❎ to start ", 1, 115, 1)
            print_centered(" press ❎ to start ", 0, 116, 7)
        end
    end
end

-- LEVELS
function update_level()
    --
    shake = max(shake - 1)
    --
    cam.x = current_runner.x - 24

    -- freeze
    if freeze_time > 0 then
        freeze_time -= 1
    else
        for o in all(objects) do
            if o.freeze > 0 then
                o.freeze -= 1
            else
                o:update()
            end

            if o.base != player and o.destroyed then
                del(objects, o)
            end
        end

        for a in all(particles) do
            a:update()
        end
    end


    if gstate != 2 then
        --end of music
        if stat(54) == -1 then
            sfx(1, 0, 8, 8)
            gstate = 2
            -- compute score
            if current_level.max_score <= 0 then -- we are not barbarians
                local max_score_counter = 0
                for o in all(objects) do
                    if o.base == leak or o.base == melon or o.base == apple or o.base == carrot then
                        max_score_counter += collected_plant_score_count
                    end
                end
                current_level.max_score = max_score_counter
            end
            current_level.best_score = max(score_count, current_level.best_score)
            local last_ratio = flr((current_level.best_score/current_level.max_score)*100)
            if last_ratio > current_level.saved_score then new_highscore = true end
            current_level.saved_score = max(last_ratio, current_level.saved_score)
            -- save high score to persistent data
            local lvl_index = find_item_table_index(current_level, level_list)
            dset(lvl_index, current_level.saved_score)
        end
    end

    if gstate == 2 then
        -- back to menu
        if btnp(❎) then
            sfx(1, 0, 16, 4)
            init_menu()
        end
        -- create fireworks
        if score_count == current_level.max_score and gtime%8 == 0 then
            spawn_particles(20, 1, cam.x + 20 + rnd(88), cam.y + 15 + rnd(40), rnd(15))
        end
    end
end

function draw_level()
    cls(1)
    
    -- camera
    if shake > 0 then
        camera(cam.x - 2 + rnd(5), cam.y - 2 + rnd(5))
    else
        camera(cam.x, cam.y)
    end

    -- draw background
    local bck_speed = 55
    local offset_x = (gtime/bck_speed) - 50
    line(cam.x - 40 - offset_x, 128, cam.x + 4 - offset_x, 35, 5)
    line(cam.x + 4 - offset_x, 35, cam.x + 68 - offset_x, 128, 5)
    line(cam.x + 30 - offset_x, 74, cam.x + 55 - offset_x, 55, 5)
    line(cam.x + 55 - offset_x, 55, cam.x + 88 - offset_x, 135, 5)
    --
    bck_speed = 4
    offset_x = (gtime/bck_speed)%256
    line(cam.x + 130 - offset_x, 128, cam.x + 150 - offset_x, 65, 5)
    line(cam.x + 150 - offset_x, 65, cam.x + 210 - offset_x, 128, 5)

    -- draw map
    map(0, 0, flr(cam.x/8)*8 , 0, 32, 16, 1)

    -- draw objects
    for o in all(objects) do
        o:draw()
    end

    for a in all(particles) do
        a:draw()
    end

    -- UI
    local last_color=peek(0x5f25)
    print("SCORE : "..score_count, cam.x + 4, cam.y + 4, 0)
    print("SCORE : "..score_count, cam.x + 3, cam.y + 3, last_color)
    print("COMBO X"..combo_count, cam.x + 4, cam.y + 12, 0)
    print("COMBO X"..combo_count, cam.x + 3, cam.y + 11, 7)

    -- End of level
    if gstate == 2 then
        print_centered("Level cleared : "..flr((score_count/current_level.max_score)*100).."%",1, cam.y + 63, 8)
        print_centered("Level cleared : "..flr((score_count/current_level.max_score)*100).."%", 0, cam.y + 62, 7)
        print_centered("best combo : "..best_combo, 1, cam.y + 71, 8)
        print_centered("best combo : "..best_combo, 0, cam.y + 70, 7)

        if new_highscore and gtime % 128 < 64 then
            print_centered("new highscore !", 1, cam.y + 86, 8)
            print_centered("new highscore !", 0, cam.y + 85, 7)
        end
        --
        print_centered("❎ back to menu", 1, 116, 0)
        print_centered("❎ back to menu", 0, 115, 7)
    end
end

-- UTILS
-- linear interpolation
function lerp(start,finish,t)
    return mid(start,start*(1-t)+finish*t,finish)
end

-- print at center
function print_centered(str, offset_x, y, col)
    print(str, cam.x + 64 - (#str * 2) + offset_x, y, col)
end

-- random range
function rrnd(min, max)
    return flr(min + rnd(max - min))
end

function find_item_table_index(item, table)
    for k, v in pairs(table) do
        if v == item then return k end
    end
    return 0
end